src/Microsoft.VisualStudio.Threading/AsyncLazy`1.cs

```diff
@@ -65,24 +75,30 @@ public AsyncLazy(Func<Task<T>> valueFactory, JoinableTaskFactory? joinableTaskFa
     /// <summary>
     /// Gets a value indicating whether the value factory has been invoked.
     /// </summary>
+    /// <remarks>
+    /// This returns <see langword="false" /> after a call to <see cref="Dispose"/>.
+    /// </remarks>
     public bool IsValueCreated
     {
         get
         {
             Interlocked.MemoryBarrier();
-            return this.valueFactory is null;
+            return this.valueFactory is null && this.value != DisposedSentinel;
```

Comment:
If I have to sprinkle any more memory barriers in here, I'm inclined to just switch to wrapping this with a lock so I have a shot at getting it right and understanding how I did it.
That said, this property is a race condition anyway. If another thread is currently in the dispose method, even if both used locks, the property will return `true` or `false` based on who wins the race. In other words, the caller can't count on the property's result outside of their own lock around all use of the AsyncLazy if they are multithreaded. So at best, this property is just a point-in-time indicator. And that said, the only thing I'm shooting for with this memory barrier is to ensure that the value returned is at least true at _some_ point in time, rather than returning some local CPU cached value that might be stale w.r.t. some other core that changed the value *seconds* ago. I'm using `Volatile.Read` for this purpose in places where I only read a single field. My hope is that a memory barrier serves the same purpose for reading multiple fields, but I'm not sure if that's true. Are you?

Reply:
The MemoryBarrier here is needed. Certainly, when IsValueCreated returns true, the value might be disposed. It is a nature race condition which cannot be prevented. I think what we should prevent is that IsValueCreated return true, while the lazy has never created any value, while another thread executing DisposeValue. It would somehow defeat the meaning of this property, so I feel we need a barrier between the two lines to set those values, even they are in a lock.